caravel/assets/javascripts/dashboard.jsx

```diff
@@ -315,17 +433,91 @@ var Dashboard = function (dashboardData) {
         }
       }
     },
+    showAddSlice: function () {
+      var slicesOnDashMap = {};
+      this.reactGridLayout.serialize().forEach(function (position) {
+        slicesOnDashMap[position.slice_id] = true;
+      }, this);
+
+      render(
+        <SliceAdder dashboard={dashboard} slicesOnDashMap={slicesOnDashMap}/>,
+        document.getElementById("add-slice-container")
+      );
+    },
+    addSlicesToDashboard: function (sliceIds) {
+      $.ajax({
+        type: "POST",
+        url: '/caravel/add_slices/' + dashboard.id + '/',
+        data: {
+          data: JSON.stringify({slice_ids: sliceIds})
+        },
+        success: function () {
+          // Refresh page to allow for slices to re-render
+          window.location.reload();
+        },
+        error: function (error) {
+          var respJSON = error.responseJSON;
+          var errorMsg = (respJSON && respJSON.message) ? respJSON.message :
+              error.responseText;
+          showModal({
+            title: "Error",
+            body: "Sorry, there was an error adding slices to this dashboard:<br />" + errorMsg
+          });
+          console.warn("Add new slices error", error);
+        }
+      });
+    },
+    saveDashboard: function () {
+      var expanded_slices = {};
+      $.each($(".slice_info"), function (i, d) {
+        var widget = $(this).parents('.widget');
+        var slice_description = widget.find('.slice_description');
+        if (slice_description.is(":visible")) {
+          expanded_slices[$(widget).attr('data-slice-id')] = true;
+        }
+      });
+      var data = {
+        positions: this.reactGridLayout.serialize(),
+        css: this.editor.getValue(),
+        expanded_slices: expanded_slices
+      };
+      $.ajax({
+        type: "POST",
+        url: '/caravel/save_dash/' + dashboard.id + '/',
+        data: {
+          data: JSON.stringify(data)
+        },
+        success: function () {
+          showModal({
+            title: "Success",
+            body: "This dashboard was saved successfully."
+          });
+        },
+        error: function (error) {
+          var respJSON = error.responseJSON;
```

Comment:
we could pull this error block out into it's own function or react component so it can be reused. 
https://en.wikipedia.org/wiki/Don%27t_repeat_yourself